nasa-gibs/worldview
# WV-2130: Implement Service Worker

## Description

This PR is to greatly simplify the PWA Service Worker to meet the base requirements, and implement it into the Progressive Web App feature.  In creating this simple service worker, I started by just having a blank JS file and using that as a service worker.  I ran the Lighthouse Audit for a PWA and was told the app was not installable because it was "Not available offline."  To get around this, I created a simple event listener for the `fetch` event, which has a callback that does nothing.  Just creating an event handler for the fetch event is enough to pass the Lighthouse Audit, it seems.  If we were to create a static offline page in the future, it would be handled in this callback.

## How To Test on Desktop
1. Clone this repo and run ` git checkout wv-2130`.
2. Run the following command:
 `npm ci && npm run build && npm start`
3. Open [localhost:3000](http://localhost:3000) in your browser.

From here, you can either install the PWA or check to see if it passes the "Lighthouse" audit feature. To install, do the following:
- In the address bar, there should be a new icon that looks like an arrow pointing down into a box.  Click that icon to install and confirm installation.
- A new icon should appear on your desktop that will launch a PWA version of Worldview.

To test via Lighthouse Audit, open the Devtools by:
- Pressing `F12` or `CTRL/CMD + SHIFT + i`.
- Clicking the "Lighthouse" tab in the Dev Tools.
- Uncheck all boxes except "Progressive Web App" and click "Generate Audit Report"
- The results should return saying the application is installable.

## How To Test on Mobile
1. Follow "How To Test On Desktop" through step 3. 
2. Find out the desktop's IP address and make sure your mobile device is connected to the same network as the desktop.
3. On mobile device, open a browser and go to:  `[desktop's IP]:3000`
_NOTE: If your request times out, your desktop may have firewall restrictions in place that block incoming connections on port 3000. Please refer to your system administrator for assistance if this is the case._
4. In your mobile browser, access where there are more options and find "Add to Homescreen".
5. This will add Worldview to your home screen which you can then tap to open as a full screen application.
